doc/developer_guide/psy_data.rst

```diff
@@ -0,0 +1,537 @@
+    provided to the wrapper library, this call will not be created
+    (and there is no need to implement this function in a wrapper
+    library).
+    
+    ``name``
+      This is the name of the variable as a string.
+    
+    ``value``
+      This is the actual content of the variable.
+    
+    The same call is used for different arguments, so a generic
+    interface is recommended to distinguish between
+    the data types provided. The netcdf kernel writer 
+    (see :ref:`user_guide:psyke_netcdf`) uses the following declaration
+    (with types defined in the dl_esm_inf library)::    
+    
+        generic, public :: PreDeclareVariable => DeclareScalarInteger, &
+                                                 DeclareScalarReal,    &
+                                                 DeclareFieldDouble
+        ...
+        subroutine DeclareScalarInteger(this, name, value)
+            implicit none
+            class(extract_PSyDataType), intent(inout), target :: this
+            character(*), intent(in) :: name
+            integer, intent(in) :: value
+        ...
+        subroutine DeclareFieldDouble(this, name, value)
+            use field_mod, only : r2d_field
+            implicit none
+            class(extract_PSyDataType), intent(inout), target :: this
+            character(*), intent(in) :: name
+            type(r2d_field), intent(in) :: value
+        ...
+        subroutine DeclareFieldDouble(this, name, value)
+            use field_mod, only : r2d_field
+            implicit none
+            class(extract_PSyDataType), intent(inout), target :: this
+            character(*), intent(in) :: name
+            type(r2d_field), intent(in) :: value
+        ...
+
+.. method:: PreEndDeclaration(this)
+
+    Called once all variables have been declared. This call is only
+    inserted if any variables are to be provided either before or after
+    the instrumented region (especially for performance profiling this
+    call is not created).
+
+.. method:: ProvideVariable(this, name, value)
+
+    This method is called for each variable to be provided to the
+    runtime library. 
+    
+    ``name``
+      This is the name of the variable as a string.
+    
+    ``value``
+      This is the actual content of the variable.
+    
+    The same method ``ProvideVariable`` is called to provide variable
+    name and content before and after the user instrumented region.
+    Again it is expected that a library using the API will provide
+    a generic interface to distinguish between the various possible data
+    types, which will be different for each infrastructure library::
+    
+        generic, public :: ProvideVariable => WriteScalarInteger, &
+                                              WriteScalarReal,    &
+                                              WriteFieldDouble
+    
+.. method:: PreEnd(this)
+
+    The method ``PreEnd`` is called after all variables before the instrumented
```

Comment:
First one done.
Second comment: yes, it is indeed not necessary, but it is highly convenient (and I admit I started without these functions): without the PreEnd and PreEndDeclaration calls you need to have a test in each PreDeclareVariable and each ProvideVariable subroutine (note that you have one of those functions for each data type - integer *4, integer *8, real, double precision, field of real, field of double precision, ... whatever) - that's just annoying to have to do that.

PreEndDeclaration is required in netcdf to switch from variable definition to data providing mode (and I need to reset the index variable used to point to the current entry in arrays). PreEnd is (from memory) not really used atm, but can be useful for the same reason (to avoid a test in each ProvideVariable subroutine). One example scenario: you want to use in-situ visualisation to look at the input fields, before you call the original code and then look at the output field. This call would give you the opportunity to view data, and click on 'ok' or so.

Reply:
Thanks for explaining.